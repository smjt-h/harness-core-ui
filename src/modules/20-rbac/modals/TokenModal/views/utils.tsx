/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getReadableDateTime } from '@common/utils/dateUtils'

export const DATE_FORMAT = 'MM/DD/YYYY'

export const hasNoExpiryDate = (value?: number): boolean => {
  return getReadableDateTime(value, DATE_FORMAT) === '12/31/2099'
}

export const getSelectedExpiration = (value?: number): string => {
  if (!value) {
    return '30'
  } else if (hasNoExpiryDate(value)) {
    return '-1'
  }
  return 'custom'
}

export const getSelectedExpirationDate = (value: string): string => {
  switch (value) {
    case '30':
    case '90':
    case '180':
      return getReadableDateTime(new Date().setDate(new Date().getDate() + parseInt(value)), DATE_FORMAT)
    case '-1':
      return '12/31/2099'
    default:
      return ''
  }
}

export const getExpirationDate = (value?: number): string => {
  if (value) {
    return getReadableDateTime(value, DATE_FORMAT)
  } else {
    return getSelectedExpirationDate(getSelectedExpiration(value))
  }
}
