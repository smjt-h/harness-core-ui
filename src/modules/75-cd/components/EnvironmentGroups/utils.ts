/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo } from 'lodash-es'
import type {
  EnvironmentGroupFilterProperties,
  EnvironmentGroupResponseDTO,
  GetEnvironmentGroupListQueryParams
} from 'services/cd-ng'

export function cleanData(values: EnvironmentGroupResponseDTO): EnvironmentGroupResponseDTO {
  const {
    name: prevName,
    identifier: prevIdentifier,
    description: prevDescription,
    tags,
    orgIdentifier,
    projectIdentifier,
    envIdentifiers
  } = values

  const name = prevName?.toString().trim()
  const identifier = prevIdentifier?.toString().trim()
  const description = prevDescription?.toString().trim()
  return {
    name,
    identifier,
    description,
    tags: defaultTo(tags, {}),
    orgIdentifier,
    projectIdentifier,
    envIdentifiers
  } as any
}

export enum SortFields {
  LastUpdatedAt = 'lastModifiedAt',
  AZ09 = 'AZ09',
  ZA90 = 'ZA90',
  Name = 'name'
}

export enum Sort {
  DESC = 'DESC',
  ASC = 'ASC'
}

export enum EnvironmentGroupDetailsTab {
  CONFIGURATION = 'CONFIGURATION',
  ENVIRONMENTS = 'ENVIRONMENTS'
}

export interface EnvironmentGroupListQueryParams
  extends Partial<
    Omit<GetEnvironmentGroupListQueryParams, 'sort' | 'getDefaultFromOtherRepo' | 'envGroupIdentifiers'>
  > {
  filters?: EnvironmentGroupFilterProperties
}
