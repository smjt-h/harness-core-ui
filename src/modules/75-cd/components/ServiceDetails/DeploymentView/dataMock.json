{
  "status": "SUCCESS",
  "data": {
    "environmentInfoByServiceId": [
      {
        "environmentName": "NewEnv",
        "environmentId": "NewEnv",
        "artifactImage": "library/nginx",
        "tag": "perl",
        "serviceName": "newService",
        "serviceId": "Serv1",
        "service_startTs": 1651059655943,
        "service_endTs": 1651059739615
      },
      {
        "environmentName": "PreProDEnv",
        "environmentId": "PreProDEnv",
        "artifactImage": "library/nginx",
        "tag": "stable",
        "serviceName": "newService",
        "serviceId": "Serv1",
        "service_startTs": 1651059739627,
        "service_endTs": 1651059835742
      },
      {
        "environmentName": "PreQa",
        "environmentId": "PreQa",
        "artifactImage": "library/nginx",
        "tag": "stable",
        "serviceName": "Serv1",
        "serviceId": "Serv1",
        "service_startTs": 1650954993799,
        "service_endTs": 1650955065955
      },
      {
        "environmentName": "env",
        "environmentId": "env",
        "artifactImage": "library/nginx",
        "tag": "stable",
        "serviceName": "Serv1",
        "serviceId": "Serv1",
        "service_startTs": 1650954674881,
        "service_endTs": 1650954993776
      }
    ]
  },
  "metaData": null,
  "correlationId": "a8e10905-f5ec-446a-8863-e0947d70b28c"
}
