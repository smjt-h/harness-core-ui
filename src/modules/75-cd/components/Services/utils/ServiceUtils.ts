/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export enum ServiceTabs {
  SUMMARY = 'summaryTab',
  Configuration = 'configuration',
  REFERENCED_BY = 'refrencedByTab',
  ActivityLog = 'activityLog'
}

const DefaultService = {
  serviceDefinition: {
    spec: {
      variables: []
    }
  }
}

export const DefaultNewStageName = 'Stage Name'
export const DefaultNewStageId = 'stage_id'

export const initialServiceState = {
  service: { ...DefaultService }
  // isLoading: false,
  // isBETemplateUpdated: false,
  // isDBInitialized: false,
  // isUpdated: false,
  // isInitialized: false,
  // gitDetails: {},
  // entityValidityDetails: {}
}
