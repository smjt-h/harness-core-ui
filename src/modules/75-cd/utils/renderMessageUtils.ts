/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { BannerType } from '@common/layouts/Constants'
import {
  getActiveUsageNumber,
  isFeatureLimitBreached,
  isFeatureOveruseActive,
  isFeatureWarningActive
} from '@common/layouts/FeatureBanner'
import type { CheckFeatureReturn } from 'framework/featureStore/featureStoreUtil'
import type { UseStringsReturn } from 'framework/strings'

interface BannerText {
  message: () => string
  bannerType: BannerType
}

export const getBannerText = (
  getString: UseStringsReturn['getString'],
  additionalLicenseProps: Record<string, unknown>,
  serviceFeatureDetail?: CheckFeatureReturn,
  dpmFeatureDetail?: CheckFeatureReturn,
  initialDeploymentsFeatureDetail?: CheckFeatureReturn
): BannerText => {
  const { isFreeEdition: isCDFree, isTeamEdition: isCDTeam } = additionalLicenseProps
  // Check for limit breach
  const isServiceLimitBreached = isFeatureLimitBreached(serviceFeatureDetail)
  const isDpmLimitBreached = isFeatureLimitBreached(dpmFeatureDetail)
  let limitBreachMessageString = ''

  if (isServiceLimitBreached && isDpmLimitBreached) {
    limitBreachMessageString = getString('cd.featureRestriction.banners.serviceAndDeploymentsLevelUp', {
      deploymentsLimit: dpmFeatureDetail?.featureDetail?.limit,
      serviceLimit: serviceFeatureDetail?.featureDetail?.limit
    })
  } else if (isServiceLimitBreached && isCDFree) {
    limitBreachMessageString = getString('cd.featureRestriction.banners.serviceLevelUp', {
      serviceLimit: serviceFeatureDetail?.featureDetail?.limit
    })
  } else if (isServiceLimitBreached && isCDTeam) {
    limitBreachMessageString = getString('cd.featureRestriction.banners.serviceLevelUpTeamEnterprise', {
      serviceLimit: serviceFeatureDetail?.featureDetail?.limit
    })
  } else if (isDpmLimitBreached) {
    limitBreachMessageString = getString('cd.featureRestriction.banners.deploymentsPerMonthLevelUp', {
      count: dpmFeatureDetail?.featureDetail?.count,
      deploymentsLimit: dpmFeatureDetail?.featureDetail?.limit
    })
  }

  if (limitBreachMessageString) {
    return {
      message: () => limitBreachMessageString,
      bannerType: BannerType.LEVEL_UP
    }
  }

  // Checking for limit usage warning
  let warningMessageString = ''
  const isServiceWarningActive = isFeatureWarningActive(serviceFeatureDetail)
  const isDpmWarningActive = isFeatureWarningActive(dpmFeatureDetail)
  const isInitialDeplWarningActive = isFeatureWarningActive(initialDeploymentsFeatureDetail)
  if (isInitialDeplWarningActive) {
    warningMessageString = getString('cd.featureRestriction.banners.initialDeploymentsWarningActive', {
      warningLimit: getActiveUsageNumber(initialDeploymentsFeatureDetail)
    })
  } else if (isServiceWarningActive) {
    warningMessageString = getString('cd.featureRestriction.banners.serviceWarningActive', {
      warningLimit: getActiveUsageNumber(serviceFeatureDetail)
    })
  } else if (isDpmWarningActive) {
    warningMessageString = getString('cd.featureRestriction.banners.dpmWarningActive', {
      count: dpmFeatureDetail?.featureDetail?.count,
      warningLimit: dpmFeatureDetail?.featureDetail?.limit
    })
  }

  if (warningMessageString) {
    return {
      message: () => warningMessageString,
      bannerType: BannerType.INFO
    }
  }

  let overuseMessageString = ''
  const isServiceOveruseActive = isFeatureOveruseActive(serviceFeatureDetail)
  if (isServiceOveruseActive && isCDTeam) {
    overuseMessageString = getString('cd.featureRestriction.banners.serviceOveruseTeamEnterprise')
  }
  if (overuseMessageString) {
    return {
      message: () => overuseMessageString,
      bannerType: BannerType.OVERUSE
    }
  }

  // If neither of limit breach/ warning/ overuse needs to be shown, return with an empty string.
  // This will ensure no banner is shown
  return {
    message: () => '',
    bannerType: BannerType.LEVEL_UP
  }
}
