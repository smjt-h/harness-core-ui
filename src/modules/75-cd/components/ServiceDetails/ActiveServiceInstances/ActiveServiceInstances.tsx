/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState, useEffect } from 'react'
import { useParams } from 'react-router-dom'
import { Card, Layout, Tab, Tabs } from '@wings-software/uicore'
import { useStrings } from 'framework/strings'
import { ActiveServiceInstancesHeader } from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstancesHeader'
import {
  GetEnvArtifactDetailsByServiceIdQueryParams,
  GetEnvBuildInstanceCountQueryParams,
  useGetEnvArtifactDetailsByServiceId,
  useGetEnvBuildInstanceCount
} from 'services/cd-ng'
import type { ProjectPathProps, ServicePathProps } from '@common/interfaces/RouteInterfaces'
import { ActiveServiceInstancesContent } from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstancesContent'
import { Deployments } from '../DeploymentView/DeploymentView'
import css from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstances.module.scss'

export enum ServiceDetailTabs {
  ACTIVE = 'Active Service Instances',
  DEPLOYMENT = 'Deployments'
}

export const ActiveServiceInstances: React.FC = () => {
  const { getString } = useStrings()
  const { accountId, orgIdentifier, projectIdentifier, serviceId } = useParams<ProjectPathProps & ServicePathProps>()
  const queryParams: GetEnvBuildInstanceCountQueryParams = {
    accountIdentifier: accountId,
    orgIdentifier,
    projectIdentifier,
    serviceId
  }

  const {
    data: activeInstancedata,
    loading: loadingActiveInstance,
    error: errorActiveInstance
  } = useGetEnvBuildInstanceCount({ queryParams })

  const queryParamsDeployments: GetEnvArtifactDetailsByServiceIdQueryParams = {
    accountIdentifier: accountId,
    orgIdentifier,
    projectIdentifier,
    serviceId
  }

  const {
    data: deploymentData,
    loading: loadingDeployment,
    error: errorDeployment
  } = useGetEnvArtifactDetailsByServiceId({ queryParams: queryParamsDeployments })

  const isDeploymentEmpty = (): boolean => {
    return Boolean(
      loadingDeployment || errorDeployment || !(deploymentData?.data?.environmentInfoByServiceId || []).length
    )
  }

  const isActiveInstanceEmpty = (): boolean => {
    return Boolean(
      loadingActiveInstance ||
        errorActiveInstance ||
        !(activeInstancedata?.data?.envBuildIdAndInstanceCountInfoList || []).length
    )
  }

  const [defaultTab, setDefaultTab] = useState(ServiceDetailTabs.ACTIVE)

  useEffect(() => {
    if (!isDeploymentEmpty() && isActiveInstanceEmpty()) setDefaultTab(ServiceDetailTabs.DEPLOYMENT)
  }, [deploymentData, activeInstancedata])

  const handleTabChange = (data: string): void => {
    setDefaultTab(data as ServiceDetailTabs)
  }

  return (
    <Card className={css.activeServiceInstances}>
      <Layout.Vertical className={css.tabsStyle}>
        <Tabs id="ServiceDetailTabs" selectedTabId={defaultTab} onChange={handleTabChange}>
          <Tab
            id={ServiceDetailTabs.ACTIVE}
            title={getString('cd.serviceDashboard.activeServiceInstancesLabel')}
            panel={
              <>
                <ActiveServiceInstancesHeader />
                <ActiveServiceInstancesContent />
              </>
            }
          />
          <Tab id={ServiceDetailTabs.DEPLOYMENT} title={getString('deploymentsText')} panel={<Deployments />} />
        </Tabs>
      </Layout.Vertical>
    </Card>
  )
}
