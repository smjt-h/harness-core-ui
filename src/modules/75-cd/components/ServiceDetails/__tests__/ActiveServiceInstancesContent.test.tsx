/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { ActiveServiceInstancesContent } from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstancesContent'
import * as cdngServices from 'services/cd-ng'
import dataMock from '../DeploymentView/dataMock.json'
import { ActiveServiceInstances } from '../ActiveServiceInstances/ActiveServiceInstances'

jest.mock('highcharts-react-official', () => () => <></>)

const mockData = {
  status: 'SUCCESS',
  data: {
    envBuildIdAndInstanceCountInfoList: [
      {
        envId: 'env1',
        envName: 'envName',
        buildIdAndInstanceCountList: [{ buildId: 'build1', count: 1 }, {}, { buildId: 'build3', count: 3 }]
      },
      {}
    ]
  }
}
jest.spyOn(cdngServices, 'useGetEnvBuildInstanceCount').mockImplementation(() => {
  return { loading: false, error: false, data: mockData, refetch: jest.fn() } as any
})
const noData = {
  status: 'SUCCESS',
  data: {}
}
describe('ActiveServiceInstancesContent', () => {
  test('should render ActiveServiceInstancesContent', () => {
    const { container, getByText } = render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <ActiveServiceInstancesContent />
      </TestWrapper>
    )
    fireEvent.click(getByText('cd.serviceDashboard.seeMore'))
    expect(container).toMatchSnapshot()
  })
  test('should render error', () => {
    jest.spyOn(cdngServices, 'useGetEnvBuildInstanceCount').mockImplementation(() => {
      return { loading: false, error: true, data: noData, refetch: jest.fn() } as any
    })
    const { container, getByText } = render(
      <TestWrapper>
        <ActiveServiceInstancesContent />
      </TestWrapper>
    )
    fireEvent.click(getByText('Retry'))
    expect(container).toMatchSnapshot()
  })

  test('should render loading', () => {
    jest.spyOn(cdngServices, 'useGetEnvBuildInstanceCount').mockImplementation(() => {
      return { loading: true, error: false, data: noData, refetch: jest.fn() } as any
    })
    const { container } = render(
      <TestWrapper>
        <ActiveServiceInstancesContent />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  test('should render no data', () => {
    jest.spyOn(cdngServices, 'useGetEnvBuildInstanceCount').mockImplementation(() => {
      return {
        loading: false,
        error: false,
        data: noData,
        refetch: jest.fn()
      } as any
    })
    const { container } = render(
      <TestWrapper>
        <ActiveServiceInstancesContent />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})

//tests differents default states of tab according to the returned api response

jest.spyOn(cdngServices, 'useGetActiveServiceInstanceSummary').mockImplementation(() => {
  return {
    loading: false,
    error: false,
    data: {},
    refetch: jest.fn()
  } as any
})
jest.spyOn(cdngServices, 'useGetInstanceGrowthTrend').mockImplementation(() => {
  return {
    loading: false,
    error: false,
    data: {},
    refetch: jest.fn()
  } as any
})
describe('ActiveInstance Tab states ', () => {
  //tab should be defaulted to deployments
  test('when activeInstance is empty and deployments is not ', () => {
    jest.spyOn(cdngServices, 'useGetEnvArtifactDetailsByServiceId').mockImplementation(() => {
      return {
        data: dataMock
      } as any
    })
    const { getByText } = render(
      <TestWrapper>
        <ActiveServiceInstances />
      </TestWrapper>
    )
    //check if table header are visible
    expect(getByText('environment')).toBeTruthy()
    expect(getByText('cd.artifactVersion')).toBeTruthy()

    //check if table rows of deployments are visible

    //value is environment column
    expect(getByText('NewEnv')).toBeTruthy()

    //corresponding artifact version
    expect(getByText('perl')).toBeTruthy()
  })

  //when both are empty then defaulted to activeInstance
  test('when both are empty ', () => {
    jest.spyOn(cdngServices, 'useGetEnvArtifactDetailsByServiceId').mockImplementation(() => {
      return {
        mutate: () => Promise.resolve({ loading: false, data: [] })
      } as any
    })
    const { getByText } = render(
      <TestWrapper>
        <ActiveServiceInstances />
      </TestWrapper>
    )
    //activeInstance tab should open
    expect(getByText('cd.serviceDashboard.noActiveServiceInstances')).toBeTruthy()
  })
})
