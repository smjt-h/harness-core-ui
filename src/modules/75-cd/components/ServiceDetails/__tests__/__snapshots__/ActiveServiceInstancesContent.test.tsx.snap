// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveServiceInstancesContent should render ActiveServiceInstancesContent 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main StyledProps--padding-top-medium"
  >
    <div
      class="TableV2--table instanceTable"
    >
      <div
        class="TableV2--header"
        role="row"
      >
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 20%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            environment
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 23%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            cd.artifactVersion
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 10%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            
          </p>
        </div>
        <div
          class="TableV2--cell"
          colspan="1"
          role="columnheader"
          style="width: 47%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            common.instanceLabel
          </p>
        </div>
      </div>
      <div
        class="TableV2--body"
      >
        <div
          class="TableV2--row TableV2--card"
          role="row"
        >
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 20%;"
          >
            <div
              class="StyledProps--font StyledProps--main envContainer"
            >
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main environmentRow StyledProps--font-size-small StyledProps--color StyledProps--color-white"
                style="--text-line-clamp: 1;"
              >
                envName
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 23%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-size-small StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-grey800"
                style="--text-line-clamp: 1;"
              >
                build1
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 10%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font StyledProps--main instanceCount overflow StyledProps--font-size-xsmall StyledProps--font-weight-bold StyledProps--background-grey100"
              >
                1
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 47%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer hexContainer StyledProps--flex StyledProps--flex-justifyContent-flex-start"
            >
              <span
                class="bp3-popover-wrapper"
              >
                <span
                  class="bp3-popover-target"
                >
                  <div
                    class="StyledProps--font StyledProps--main hex StyledProps--background-primary3 StyledProps--margin-left-xsmall StyledProps--margin-right-xsmall StyledProps--margin-top-xsmall StyledProps--margin-bottom-xsmall"
                    style="width: 18px; height: 18px;"
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="TableV2--row TableV2--card"
          role="row"
        >
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 20%;"
          />
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 23%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font StyledProps--main seeMore StyledProps--font-size-xsmall StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-primary6 StyledProps--background-primary1"
              >
                cd.serviceDashboard.seeMore
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 10%;"
          />
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 47%;"
          />
        </div>
        <div
          class="TableV2--row TableV2--card"
          role="row"
        >
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 20%;"
          />
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 23%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-size-small StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-grey800"
                style="--text-line-clamp: 1;"
              >
                build3
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 10%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font StyledProps--main instanceCount overflow StyledProps--font-size-xsmall StyledProps--font-weight-bold StyledProps--background-grey100"
              >
                3
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 47%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer hexContainer StyledProps--flex StyledProps--flex-justifyContent-flex-start"
            >
              <span
                class="bp3-popover-wrapper"
              >
                <span
                  class="bp3-popover-target"
                >
                  <div
                    class="StyledProps--font StyledProps--main hex StyledProps--background-primary3 StyledProps--margin-left-xsmall StyledProps--margin-right-xsmall StyledProps--margin-top-xsmall StyledProps--margin-bottom-xsmall"
                    style="width: 18px; height: 18px;"
                  />
                </span>
              </span>
              <span
                class="bp3-popover-wrapper"
              >
                <span
                  class="bp3-popover-target"
                >
                  <div
                    class="StyledProps--font StyledProps--main hex StyledProps--background-primary3 StyledProps--margin-left-xsmall StyledProps--margin-right-xsmall StyledProps--margin-top-xsmall StyledProps--margin-bottom-xsmall"
                    style="width: 18px; height: 18px;"
                  />
                </span>
              </span>
              <span
                class="bp3-popover-wrapper"
              >
                <span
                  class="bp3-popover-target"
                >
                  <div
                    class="StyledProps--font StyledProps--main hex StyledProps--background-primary3 StyledProps--margin-left-xsmall StyledProps--margin-right-xsmall StyledProps--margin-top-xsmall StyledProps--margin-bottom-xsmall"
                    style="width: 18px; height: 18px;"
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActiveServiceInstancesContent should render error 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main"
    data-test="ActiveServiceInstancesError"
    style="height: 360px;"
  >
    <div
      class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
      style="width: 100%; height: 100%;"
    >
      <div
        class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
        style="align-items: center; width: 500px;"
      >
        <span
          class="bp3-icon bp3-icon-error StyledProps--main StyledProps--color StyledProps--color-red500"
          icon="error"
        >
          <svg
            data-icon="error"
            height="32"
            viewBox="0 0 20 20"
            width="32"
          >
            <desc>
              error
            </desc>
            <path
              d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-align-center StyledProps--color StyledProps--color-red500"
        >
          We cannot perform your request at the moment. Please try again.
        </p>
        <button
          class="bp3-button bp3-disabled bp3-loading Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-primary"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <div
            class="bp3-spinner bp3-button-spinner"
          >
            <div
              class="bp3-spinner-animation"
            >
              <svg
                height="20"
                stroke-width="16.00"
                viewBox="-3.00 -3.00 106.00 106.00"
                width="20"
              >
                <path
                  class="bp3-spinner-track"
                  d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                />
                <path
                  class="bp3-spinner-head"
                  d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                  pathLength="280"
                  stroke-dasharray="280 280"
                  stroke-dashoffset="210"
                />
              </svg>
            </div>
          </div>
          <span
            class="bp3-button-text"
          >
            Retry
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActiveServiceInstancesContent should render loading 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main"
    data-test="ActiveServiceInstancesLoader"
    style="height: 360px;"
  >
    <div
      class="StyledProps--font StyledProps--main PageSpinner--spinner StyledProps--flex StyledProps--flex-align-center-center"
    >
      <div
        class="Layout--vertical Layout--layout-spacing-medium StyledProps--main PageSpinner--content"
        style="align-items: center; width: 500px;"
      >
        <span
          class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey600"
          data-icon="steps-spinner"
        >
          <svg
            height="32"
            viewBox="0 0 128 128"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M122.5 69.25H96.47a33.1 33.1 0 000-10.5h26.05a5.25 5.25 0 010 10.5z"
                fill="#ff8e02"
              />
              <path
                d="M112.04 97.83L89.47 84.8a33.1 33.1 0 005.25-9.1l22.57 13.03a5.25 5.25 0 01-5.28 9.1zM88.68 117.35L75.65 94.78a33.1 33.1 0 009.1-5.25l13.02 22.57a5.25 5.25 0 11-9.1 5.25zM58.7 122.57V96.5a33.1 33.1 0 0010.5 0v26.07a5.25 5.25 0 01-10.5 0zM30.1 112.1l13.04-22.57a33.1 33.1 0 009.1 5.25L39.2 117.35a5.25 5.25 0 11-9.1-5.25zM10.6 88.74L33.16 75.7a33.1 33.1 0 005.25 9.1L15.88 97.83a5.25 5.25 0 11-5.25-9.1z"
                fill="#ffddb3"
                fill-opacity="0.3"
              />
              <path
                d="M5.37 58.75h26.06a33.1 33.1 0 000 10.5H5.37a5.25 5.25 0 010-10.5z"
                fill="#ffd29a"
                fill-opacity="0.4"
              />
              <path
                d="M15.85 30.17L38.4 43.2a33.1 33.1 0 00-5.24 9.1L10.6 39.25a5.25 5.25 0 115.25-9.1z"
                fill="#ffc680"
                fill-opacity="0.5"
              />
              <path
                d="M39.2 10.65l13.03 22.57a33.1 33.1 0 00-9.1 5.25l-13-22.57a5.25 5.25 0 119.1-5.25z"
                fill="#ffbb67"
                fill-opacity="0.6"
              />
              <path
                d="M69.2 5.43V31.5a33.1 33.1 0 00-10.5 0V5.42a5.25 5.25 0 1110.5 0z"
                fill="#ffb04d"
                fill-opacity="0.7"
              />
              <path
                d="M97.77 15.9L84.75 38.47a33.1 33.1 0 00-9.1-5.25l13.03-22.57a5.25 5.25 0 119.1 5.25z"
                fill="#ffa535"
                fill-opacity="0.8"
              />
              <path
                d="M117.3 39.26L94.7 52.3a33.1 33.1 0 00-5.25-9.1l22.57-13.03a5.25 5.25 0 015.25 9.1z"
                fill="#ff991b"
                fill-opacity="0.9"
              />
              <animatetransform
                attributeName="transform"
                calcMode="discrete"
                dur="1200ms"
                repeatCount="indefinite"
                type="rotate"
                values="0 64 64;30 64 64;60 64 64;90 64 64;120 64 64;150 64 64;180 64 64;210 64 64;240 64 64;270 64 64;300 64 64;330 64 64"
              />
            </g>
          </svg>
        </span>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-size-medium StyledProps--font-align-center StyledProps--color StyledProps--color-grey600"
        >
          Loading, please wait...
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActiveServiceInstancesContent should render no data 1`] = `
<div>
  <div
    class="Layout--vertical StyledProps--main activeServiceInstancesEmpty StyledProps--flex StyledProps--flex-align-center-center"
    data-test="ActiveServiceInstancesEmpty"
    style="height: 360px;"
  >
    <div
      class="StyledProps--font StyledProps--main StyledProps--margin-bottom-medium"
    >
      <img
        height="50"
        src="test-file-stub"
        style="align-self: center;"
        width="50"
      />
    </div>
    <p
      class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
    >
      cd.serviceDashboard.noActiveServiceInstances
    </p>
  </div>
</div>
`;
